File: tests/fixtures/python/safe.py
import secrets
import subprocess

# Safe SQL - parameterized query
def login(username, password):
    cursor.execute("SELECT * FROM users WHERE username=? AND password=?", (username, password))

# Safe subprocess - no shell
def run_command(args):
    subprocess.run(["ping", "-c", "1", args], shell=False)

# Safe random
def generate_token():
    return secrets.token_hex(32)

# Safe file access with validation
def download_file(filename):
    import os
    safe_path = os.path.join("/uploads", os.path.basename(filename))
    if not safe_path.startswith("/uploads"):
        raise ValueError("Invalid path")
    return open(safe_path)

File: tests/fixtures/javascript/safe.js
const crypto = require('crypto');
const path = require('path');

// Safe - textContent instead of innerHTML
function renderComment(comment) {
    document.getElementById('comments').textContent = comment;
}

// Safe random
function generateId() {
    return crypto.randomUUID();
}

// Safe path handling
function readUserFile(req, res) {
    const filename = path.basename(req.query.filename);
    const filePath = path.join('/uploads', filename);
    if (!filePath.startsWith('/uploads')) {
        return res.status(403).send('Forbidden');
    }
    fs.readFile(filePath, (err, data) => {
        res.send(data);
    });
}

// Safe - parameterized MongoDB query
async function findUser(req, res) {
    const username = String(req.body.username);
    const user = await db.collection('users').findOne({ username });
    res.json(user);
}
